### PR TITLE
[[FIX]] Enforce TDZ within initializer of lexical declaration

### LINKED ISSUE #2637: TDZ is not enforced within initializer of lexical declaration
The following code should produce [error E056](https://github.com/jshint/jshint/blob/18f8775a05c10fc8c157ef7b79c16d36ba0bbc7b/src/messages.js#L74) but currently does not:

```js
// jshint esversion: 6
let a = a;
const b = b;
```

See also: gh-2632

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
Fixes #2637

### CODE REVIEW

File: tests/unit/core.js [hunk 1]
```diff
@@ -1989,3 +1989,42 @@ exports.duplicateProto = function (test) {
...
+    "};",
+    "const f = {",
+    "  x: f,",
+    "  y: () => f",
+    "};",
+    // line 13
+    "let g, h = g;",
+    "const i = 0, j = i;",
+    "let [ k, l ] = l;",
+    "const [ m, n ] = n;",
```
Comment (nicolo-ribaudo) [inline tests/unit/core.js]: Is it correct to throw on lines 15 and 16? (Babel doesn't)

**Comment by nicolo-ribaudo** [comment]
Just added other TDZ checks

**Comment by jugglinmike** [comment]
Thanks for the patch, @nicolo-ribaudo! I have a different approach in mind--could you tell me what you think of my changes in gh-2733?

**Comment by nicolo-ribaudo** [comment]
Your approach looks better than mine, I'm closing this pr.

**Pull request closed without merging**